File: scripts/normalize_name.js
/**
 * Vereinheitlichte Namensfunktion für KPI-Dateien (Frontend).
 * Diese Version ist 1:1 identisch mit normalize_name.py (Backend).
 * Sie sorgt dafür, dass Titel aus available_kpis.json exakt dieselben Dateinamen ergeben
 * wie die vom Backend erzeugten Dateien in /data/.
 */
window.normalizeName = function(title) {
  if (!title) return "unknown_kpi";

  let s = title.toLowerCase().trim();

  // Subskript-Zahlen vereinheitlichen
  const subDigits = { "₀":"0", "₁":"1", "₂":"2", "₃":"3", "₄":"4", "₅":"5", "₆":"6", "₇":"7", "₈":"8", "₉":"9" };
  s = s.replace(/[₀₁₂₃₄₅₆₇₈₉]/g, m => subDigits[m]);

  // CO₂ → co2
  s = s.replace("co₂", "co2");

  // "(% of gdp)" oder "percent of gdp" → "of gdp"
  s = s.replace(/\s*(%|percent)\s*of\s*gdp/, " of gdp");

  // Punkt zwischen Ziffern (2.5) → Unterstrich
  s = s.replace(/(?<=\d)\.(?=\d)/g, "_");

  // "1,000" oder "1.000" → "1_000"
  s = s.replace(/1[.,]000/g, "1_000");

  // Unnötige Zeichen entfernen
  s = s.replace(/[(),]/g, "");

  // Nicht alphanumerische Zeichen → Unterstrich
  s = s.replace(/[^a-z0-9._-]/g, "_");

  // Bindestriche → Unterstriche
  s = s.replace(/-/g, "_");

  // Doppelte Unterstriche reduzieren
  s = s.replace(/_+/g, "_");

  // Führende und abschließende Unterstriche entfernen
  s = s.replace(/^_+|_+$/g, "");

  return s;
};

File: scripts/normalize_name.py
import re

def normalize_name(title: str) -> str:
    """
    Einheitliche Normalisierung für KPI-Dateien (Backend & Frontend identisch).
    Regeln (müssen 1:1 mit normalize_name.js übereinstimmen):
      - Kleinschreibung, Trim
      - Subskript-Ziffern ₀–₉ -> 0–9
      - "co₂" -> "co2"
      - "% of gdp" / "percent of gdp" -> "of gdp"
      - Zahlen mit Punkt (z. B. 2.5) -> 2_5
      - "1,000" oder "1.000" -> "1_000"
      - Nicht alphanumerische Zeichen -> "_"
      - Bindestriche -> "_"
      - Mehrfache "_" reduzieren, führende/Trailing "_" entfernen
    """
    if not title:
        return "unknown_kpi"

    s = title.lower().strip()

    # Subskript-Zahlen vereinheitlichen
    sub_map = str.maketrans("₀₁₂₃₄₅₆₇₈₉", "0123456789")
    s = s.translate(sub_map)

    # Ersetzungen/Korrekturen
    s = s.replace("co₂", "co2")

    # "(% of gdp)" / "percent of gdp" -> "of gdp"
    s = re.sub(r'\s*(%|percent)\s*of\s*gdp', ' of gdp', s)

    # Punkt zwischen Ziffern (2.5) -> Unterstrich
    s = re.sub(r'(?<=\d)\.(?=\d)', '_', s)

    # "1,000" oder "1.000" -> "1_000"
    s = re.sub(r'1[.,]000', '1_000', s)

    # Unnötige Zeichen entfernen
    s = s.replace("(", "").replace(")", "").replace(",", "")

    # Nicht alphanumerische Zeichen -> "_"
    s = re.sub(r'[^a-z0-9._-]', '_', s)

    # Bindestriche -> Unterstriche
    s = s.replace("-", "_")

    # Doppelte Unterstriche reduzieren
    s = re.sub(r'_+', '_', s)

    # Führende/trailing "_" entfernen
    s = s.strip('_')

    return s
